### PR TITLE
MediaTek 8195 Audio DSP, take 2

### PR DESCRIPTION
**Pull request by andyross**
This is getting hot again.  Rebase, update for changes and push again.  Will split out the early patches as separate PRs while this is in review.

### CODE REVIEW

**Comment by andyross** [comment]
Split-out platform-layer PRs in #72768 #72769 and #72771 

Please review that content there, I'll remove from here as they merge.

File: drivers/interrupt_controller/intc_mtk_adsp.c [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+struct intc_mtk_cfg {
+	uint32_t xtensa_irq;
+	uint32_t irq_mask;
+	uint32_t sw_isr_off;
+	volatile uint32_t *enable_reg;
+	volatile uint32_t *status_reg;
+};
+
+bool intc_mtk_adsp_get_enable(const struct device *dev, int irq)
```
Comment (cfriedt) [inline drivers/interrupt_controller/intc_mtk_adsp.c:17]: There might be a reason this is in the global scope, but if not, I would suggest keeping this and other functions below static.
```suggestion
static bool intc_mtk_adsp_get_enable(const struct device *dev, int irq)
```

**Review by cfriedt — DISMISSED** [review]
LGTM other than maybe a missing `static` qualifier.

File: drivers/interrupt_controller/intc_mtk_adsp.c (hunk 1 above)
Comment (andyross) [reply drivers/interrupt_controller/intc_mtk_adsp.c:17]: Yeah, these need to be called out of the SOC-side irq_en/disable() implementation (in soc/.../irq.c in this PR).  There's no generic mechanism for that.  There was some discussion about a new API (c.f. @ycsin @bjarki-trackunit) a few months back that would tie things together, but it's a big task and everyone has different requirements.

**Review by dcpleung — DISMISSED** [review]

**Comment by andyross** [comment]
Stick a DNM on temporarily now that I see this has some +1's just so it doesn't merge before the dependee PRs

**Comment by andyross** [comment]
Noting here also: #72771 and #72768 still need some review love.

**Comment by andyross** [comment]
Spun on kconfig linting a bit, should be good now.  The two pylint warnings are false positives: in that particular construct an iterate has no advantages.

**Review by cfriedt — APPROVED** [review]
Refresh +1

**Review by dcpleung — APPROVED** [review]

**Comment by andyross** [comment]
Ping for merge?  As mentioned, the pylint warning is contra-idiomatic (this is a bytewise copy over an mmap region, doing it with python enumerate() would be needlessly confusing).  Also there's a doc build failure which seems to be a CI glitch?

**Review by ceolin — APPROVED** [review]

**Review by peter-mitsis — APPROVED** [review]

**Comment by andyross** [comment]
To clarify the pylint issue, which provoked some discussion on discord, pylint has a test where it detects where you are doing "`for i in range(len(some_iterable))` and using both the index `i` and the element `some_iterable[i]`.  And it suggests that you use `enumerate()` instead, which returns both.

Which is all well and good for the typical lists and dics with which the authors assumed it would be used.  But in the cases it's flagging I'm iterating into **two** arrays (bytes and mmap objects) with the same index, and the asymmetry is confusing and weird.  It wants to turn a trivially obvious bytewise copy that looks like this:

```python
for i in range(len(input)):
    output[i] = input[i]
```

into this asymmetric silliness, which obscures the actual operation that's happening:

```python
for i, elem in enumerate(input):
    output[i] = elem
```

It's just an inappropriate warning.  In fact it's not even at "[W]arning" level, it's a "[C]onvention" message, asking me to "consider" using the alternative construction.  And I did consider it!   And I won't, because it's silly.

**Comment by cfriedt** [comment]
Maybe @nashif?

**Comment by nashif** [comment]
> Ping for merge? As mentioned, the pylint warning is contra-idiomatic (this is a bytewise copy over an mmap region, doing it with python enumerate() would be needlessly confusing). Also there's a doc build failure which seems to be a CI glitch?

you can add 

```
# pylint: disable=consider-using-enumerate
``` 

and just ignore if you think it is a false positive. This is the best way to deal with something like that, not by disabling the rule altogethe. We had this rule for years and in some cases it makes sense.

**Pull request merged**